pyfr/plugins/acoustics.py

```diff
@@ -0,0 +1,2106 @@
+            F -= np.mean(F, 0)
+            Q = np.einsum('i,ij->ij', self.wwind, Q)
+            F = np.einsum('i,ijk->ijk', self.wwind, F)
+
+        # perform fft of Q, F fluxes
+        Qfft = np.empty((self.nfreq, self.nqpts), dtype=np.complex64)
+        Ffft = np.empty((self.nfreq, self.nqpts, self.ndims), 
+                                                dtype=np.complex64)
+        F = np.swapaxes(F, 0, 2).reshape(-1, self.ntsub) #shape:nt,nqpts,ndim
+        Qfft = realfft(Q.T).T
+        Ffft = realfft(F).reshape(self.ndims, self.nqpts, -1)
+        Ffft = np.swapaxes(Ffft, 0, 2)
+
+        #compute pfft, i=nob, j=nfreq, k=nqpts, p shape i,j,k 
+        kwvR = np.einsum('ik,j->jik', mR, kwv)
+        exp_term0 = np.exp(-1j*kwvR)        # exp(-ikR)
+        exp_term1 = exp_term0/mRs          # exp(-ikR)/R*
+        exp_term2 = exp_term0/(mRs*mRs) # exp(-ikR)/(R* x R*)
+        # p1_term
+        pfreq  = 1j*np.einsum('j,jik,jk->ijk', omega, exp_term1, Qfft)
+        # p2_term0 
+        pfreq += 1j*np.einsum('j,jik,ikm,jkm->ijk',
+                                        kwv, exp_term1, nR, Ffft)   
+        # p2_term1 
+        pfreq += np.einsum('jik,ikm,jkm->ijk', exp_term2, nRs, Ffft) 
+
+        #surface integration
+        #i=nobservers, j=nfreq or ntime, k=nqpts
+        pfft =  np.einsum('ijk,k->ij', pfreq, self.qdA)
+        pfft /= (4.*np.pi)
+        if self.window:
+            pfft *= self.windscale
+        
+        return pfft
+
+
+class PointAcousticSrc(object):
+    def __init__(self, name, srclocs, flowinfo, *srcdata):
+        self.name = name
+        self.srclocs = srclocs
+        self.ndims = len(srclocs)
+        self.nvars = len(flowinfo)
+        self.rhoinf = flowinfo[0]
+        self.uinf = flowinfo[1:self.ndims+1]
+        self.pinf = flowinfo[-1]
+
+    def acoustic_psoln(self, *args):
+        pass
+    def update_usoln(self, *args):
+        pass
+
+class MonopoleSrc(PointAcousticSrc):
+    def __init__(self, name, srclocs, flowinfo, ampl, srcfreq, ntime, nperiods,
+                 gamma):
+        super().__init__(name, srclocs, flowinfo, ampl, srcfreq, ntime,
+                         nperiods, gamma)
+        self.gamma = gamma
+        self.co = np.sqrt(self.gamma*self.pinf/self.rhoinf)
+        self.ampl = ampl
+        self.srcfreq = srcfreq
+        self.nperiods = nperiods
+        self.nt = ntime
+        self.omega = 2.*np.pi*self.srcfreq
+        self.tperiod = 2.*np.pi/self.omega
+        self.dt = self.nperiods*self.tperiod/self.nt
+        self.freq = np.fft.rfftfreq(self.nt,self.dt)
+        self.nfreq = len(self.freq)
+        
+    def acoustic_psoln(self, xyz_ob):
+        nobserv = xyz_ob.shape[0]
+        ptime = np.empty((self.nt, nobserv))
+        pfft = np.empty((nobserv, self.nfreq), dtype=np.complex64)
+
+        tarr = np.arange(0, self.nt)*self.dt
+        for i, ti in enumerate(tarr):
+            ptime[i] = self._comput_exact_psoln(xyz_ob, ti, ptime[i])
```

Comment:
This smells funny (mutable on LHS and RHS).

Reply:
can be improved for sure.